### DISCUSSION TITLE
ChatGPT is significantly superior to copilot, when will it get upgraded ?

### DISCUSSION THREAD

**Discussion by cmp-nct** [community/community#41551] [Copilot Conversations]
### Select Topic Area

Product Feedback

### Body

I've been comparing the paid copilot and the free chatgpt for code generation and the comparison is like an adult with a toddler.

Copilot generates bad code with rare nice snippets, it is only good when looking at very localized code and small snippets.
ChatGPT on the other hand generates fully functional code and can even modify it afterwards.

Copilot only has 1500 tokens it can "see", which is a tiny fraction of the whole code it is supposed to work with.
Even the code copilot generates can get so large that it starts to forget what it is even creating.

ChatGPT is 4500 tokens and I've not seen it generating garbage, Copilot generates garbage frequently.

**Comment by MulleDK19** [comment]
> I've been comparing the paid copilot and the free chatgpt for code generation and the comparison is like an adult with a toddler.

What have you tried, specifically?


> Copilot only has 1500 tokens it can "see", which is a tiny fraction of the whole code it is supposed to work with.
> Even the code copilot generates can get so large that it starts to forget what it is even creating.

Don't know where you're getting that from. Copilot uses Codex behind the scenes which supports 8,000 tokens, which is around 1600 **words**.

**Reply by cmp-nct** [reply]
So I mixed up tokens and words, doesn't change what I said in any way.

I've tried hundreds of codes on ChatGPT and thousands of codes in Copilot.
The quality difference is like comparing a toddler with a student, given that the toddler might just start drawing the same symbol over and over that's quite fitting.

ChatGPT is not just an "upgrade" of Codex to transform it into a chat-type interface, it is an extreme quality improvement.
The code it produces is almost always coherent, high quality and it has not a single time produced "garbage".
It's not always working code but the difference in quality is extreme.


For work I started using more chat GPT than copilot, despite the fact that it's slower and not integrated into the IDE and requires a specialized prompt for each task.
The reason is that it typically generates something I can use.
I only use copilot for copy/paste jobs.

Copilot is nice if you have something that is repeatable work or if you are writing a few branches which are similar.
It is mostly a completion agent that can complete stuff the usual completion can't.
ChatGPT can actually understand and write code.

**Reply by KiameV** [reply]
I have to agree here. I've been using ChatGPT more and more. I've found it's limitations - it's naive regarding public API's (assuming certain REST structures) and makes some assumptions about how libraries work - but the code it writes is easy to use, mold, and change. It keeps context too so once it's done writing, I can ask it to modify what it just wrote.

ChatGPT has been able to write loads of functions, some highly complex, without any modifications to fit into my code. Even when it does make mistakes, they're easy to find and fix.

If copilot was on par, or hopefully better, than ChatGPT, I'd likely go from a programmer to a code integrator - having copilot write my code and I figuring out how all the code fits together.

**Comment by KiameV** [comment]
The fact that ChatGPT can seamlessly write great golang unit tests for my code is amazing. If copilot could start doing that, that would be amazing as ChatGPT is annoying to interface with.

**Reply by dawoodjee** [reply]
I agree, an upgrade would be a no brainer

**Reply by bomzj** [reply]
I agree 🚀 ChatGpt is  far superior to Copilot and other paid alike tools. There is no point to waste time on Copilot when we have ChatGpt.

**Reply by tmlewin** [reply]
> I agree 🚀 ChatGpt is far superior to Copilot and other paid alike tools. There is no point to waste time on Copilot when we have ChatGpt.

I have terminated my copilot subscription. it was awesome when it kicked off but the code and suggestion quality has not improved at all. I have complained and the customer service doesnt seem to care . chatgpt has way better code quality

**Reply by cmp-nct** [reply]
> > I agree 🚀 ChatGpt is far superior to Copilot and other paid alike tools. There is no point to waste time on Copilot when we have ChatGpt.
> 
> I have terminated my copilot subscription. it was awesome when it kicked off but the code and suggestion quality has not improved at all. I have complained and the customer service doesnt seem to care . chatgpt has way better code quality

Though the quality did improve recently (and degrade at the same time).
On the up side:
They now take a few lines below the caret into account, so the code quality is a lot better.
They also seem to add something from other files, I'm not sure what but they add some context which is better than nothing.

On the downside:
The vast majority of recommendations is now one line only, they do not seem to understand that people sometimes need one word, one line or one page of code so they opted for the cheapest solution.
The functionality to get "more" (backward,forward)  suggestions is disabled, I've not had a single instance where it worked anymore.

Of course Copilot it is still based on davinci 002 and not davinci 003, "copilot/codex" is far behind in code quality.
I've quite some experience with the GPT models they offer and softly said: they are all mental retards except for davinci-003. Chat GPT is a modified 003

**Reply by MulleDK19** [reply]
CoPilot is NOT Davinci-002. It's not Davinci at all. Davinci is 175 billion parameters, Copilot is 13 billion parameters.

**Reply by cmp-nct** [reply]
You should not "affirm" your beliefs so strongly without actually informing yourself properly on a subject.
So short said: you are wrong
![image](https://user-images.githubusercontent.com/78893154/216634015-c3dc1ea7-425a-4487-8069-62af2a60adc2.png)

**Reply by MulleDK19** [reply]
No, that is the copilot model.. which is 13 billion.... Davinci is and always has been 175, so copilot isn't "still" based on davinci-2. Code-davinci is not davinci.

**Comment by adamsg1** [comment]
I compared ChatGPT to copilot with C++. For the relatively simple mathematical function I asked for, ChatGPT gave an accurate answer, and a useful unit test that asserts on the correctness of the response. CoPilot forgot consts, references, missed edge cases in the algorithm, and provided a useless "unit test" that simply called the function and printed the outputs. 
I guess copilot would have been considered impressive before ChatGPT, but now that I know what ChatGPT can do (when it has a few spare cycles to respond to my prompts) it is hard to continue to use copilot. 
I am ending my 2 free months soon, and if nothing changes, I am out...

**Reply by mpottinger** [reply]
We need IDE integration. I would love GPT-4 based completions right in the IDE Copilot style.

**Comment by fritzstauff** [comment]
Any news on this? Would be nice to have gpt-3 or 4 level that understands our code base.

**Reply by mpottinger** [reply]
I've been thinking of rolling my own using the API, including condensed summaries of my project code as part of its prompting.
